enhancements/microshift/microshift-storage-version-migrator.md

```diff
@@ -0,0 +1,267 @@
+
+1. A risk that we have with this feature is that currently there is no real
+   upstream solution for storage version migrations. The current implementation
+   is in alpha and it does seem like it will be revisited in the future.
+   Luckily, until there is more consensus upstream we can continue to use the
+   approach outlined here but we will need to keep in mind that we might need to
+   re-work this in the future.
+
+2. In that same vein, we do run the risk that the code doesn't get touched much
+   due to the nature and life of the current Storage Version Migrator solution.
+   So we might run into dependency problems in the future. The good news there
+   is that we have good ties to the upstream developers so the procedure for
+   resolving any future dependency problems would be the same as the other
+   services we run.
+
+3. Another risk is that the new [Aggregated
+   Discovery](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3352-aggregated-discovery)
+   removes the `StorageVersionHash`, luckily the change does take into account
+   that the Migrator uses it and we just need to make sure our [discovery client
+   is in legacy
+   mode](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3352-aggregated-discovery#is-the-rollout-accompanied-by-any-deprecations-andor-removals-of-features-apis-fields-of-api-types-flags-etc)
+
+4. Lastly, migrations can fail, so there can be situations when a migration
+   might fail on a specific resource. We will be able to catch these failures in
+   our own e2e tests for our resources. However, when it comes to user
+   resources, it is up to the users to update those resources to resolve the
+   failure.
+
+### Drawbacks
+
+The technical aspect of this addition is relatively simple, some drawbacks that
+we face are that we now have two services added to our code base. This does mean
+a small increase in CPU usage by the core MicroShift binary and more code for us
+to keep track of.
+
+## Design Details
+
+### Open Questions
+
+In our look at this tool we found out that we don't run the trigger controller
+in OpenShift due to the potential problems that might arise from mass updates of
+large clusters and the traffic load on API Server during migrations. This really
+isn't as much of a concern for MicroShift since we don't run at the same scale
+as OpenShift, and MicroShift is designed to operate with a smaller API
+footprint. Queries are generally smaller and being at the edge means that the
+use case isn't a typical HA deployment where multiple teams are working on multi
+node clusters, so an update to a resource does not have the same potential to
+touch thousands of resources. Since it is the case that we run with a smaller
+embedded API footprint, do we need to run the trigger logic at all or is it
+worth MicroShift shipping `StorageVersionMigration` with upgrades when they are
+deemed to be needed for embedded components?
```

Comment:
The trigger API tracks the storage version, right? So it only actually does work if that changes?

What is the cost of always doing that check vs. the work to not always do it?

Reply:
That's right, the trigger will query for the API Resources, and the `StorageState` for that specific API Resource, and if there is a diff on the `StorageVersionHash` or it's stale (meaning the trigger crashed and the heartbeat is older than 20 minutes), it'll create a `StorageVersionMigration` for the migrator to act on.

After the initial creation, it'll be a discovery query every 10 minutes, and `StorageState` check and update per API Discovery resource. I haven't measured the CPU hit for that, but it would be fair to say we'll see a rhythmic check on the API Server every 10 minutes.

Reply:
Based on the conversation in https://github.com/openshift/enhancements/pull/1445/files#r1276763783 I don't expect many storage changes to happen. Maybe we can avoid running the trigger at all for now and describe how to use the API when an update is known to be necessary. If that's not sufficient, we could look at running the logic on startup, which is when we do the other work related to upgrades.

Reply:
I have no concerns doing either one of those, giving guidance would be preferred since it's more inline with the current OpenShfit approach. 

If we still decide to just run the trigger logic, as long as we keep it simple and just do a regular discovery and creation of `StorageVersionMigration` then that would be pretty straight forward. Just as long as we don't try to go back to the original idea of having this run in a pre-run state where only the API Server and ETCD are up, we should be good to make that change with out much problem.

Reply:
Right, I was thinking we would just run the trigger logic on startup but not at 10 minute intervals. We don't expect resources to change that much on a given instance.

Let's turn the trigger off entirely for now, but leave it there to make it easy to restore if we decide we do need it.

Reply:
We discussed that change in slack and agreed to continue with this PR describing how things actually work and come back and update it if we disable the trigger.